NREL bird clearsky

mikofski opened the pull request:
Yet another clear sky model, based on the [MS Excel workbook by NREL Daryl Myers][1] and the model described in the paper ["A Simplified Clear Sky model for Direct and Diffuse
Insolation on Horizontal Surfaces"][2] by R.E. Bird and R.L Hulstrom, SERI Technical
Report SERI/TR-642-761, Feb 1981. Solar Energy Research Institute, Golden, CO.

![bird_clearsky_model](https://cloud.githubusercontent.com/assets/1385621/20858661/4814a3c8-b8ff-11e6-9e5c-bfd4ea7d8050.png)

[1]: http://rredc.nrel.gov/solar/models/clearsky/
[2]: http://rredc.nrel.gov/solar/pubs/pdfs/tr-642-761.pdf

wholmgren commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+    )
+    gh = np.where(airmass > 0, (id_nh + ias) / (1.0 - alb * rs), 0.0)
+    testvalues = (day_angle, declination, eqt, hour_angle, zenith, airmass)
+    return id_, id_nh, gh, gh - id_nh, testvalues
```
Better to do the subtraction on a separate line.

mikofski replied:
okay, changed to `diffuse_horiz`
:heavy_check_mark:

wholmgren commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+        Solar Radiation Model 
+From the publication "A Simplified Clear Sky model for Direct and Diffuse
+Insolation on Horizontal Surfaces" by R.E. Bird and R.L Hulstrom, SERI Technical
+Report SERI/TR-642-761, Feb 1991. Solar Energy Research Institute, Golden, CO.
```
1981

mikofski replied:
:heavy_check_mark:

wholmgren commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+Insolation on Horizontal Surfaces" by R.E. Bird and R.L Hulstrom, SERI Technical
+Report SERI/TR-642-761, Feb 1991. Solar Energy Research Institute, Golden, CO.
+
+The model is based on comparisons with results from rigourous radiative transfer
```
text needs to be edited so that it's not specific to the excel file. Not sure if there is a licensing/attribution issue here.

mikofski replied:
:heavy_check_mark:

wholmgren commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+
+def bird(doy, hr, lat, lon, tz, press_mB, o3_cm, h2o_cm, aod_500nm, aod_380nm,
+         b_a, alb, lyear=365.0, solar_constant=1367.0):
+    """
```
Some (all?) of the edited doc string above should be moved here.

mikofski replied:
:heavy_check_mark:

jforbess commented:
Also not a fan of a module called nrel.py for this purpose, and I think
that the bird model should be included in clearsky.py. Not sure about the
new package concept for clearsky.py, but if it is, would prefer the module
name to be bird or include bird.

On Sat, Dec 3, 2016 at 1:43 PM, Will Holmgren <notifications@github.com>
wrote:

> *@wholmgren* commented on this pull request.
>
> Thanks! I think the Bird model would be a useful addition, though I admit
> I don't have any experience with it.
>
> I'll put the API-level comments here. The resolution of these might make
> some of my line-by-line comments irrelevant.
>
>    1. Restructuring clearsky.py into a package: maybe we should do this
>    first in a separate PR? On a related note, I thought you were an proponent
>    of api.py in #235 <https://github.com/pvlib/pvlib-python/issues/235>
>    so I was a little surprised that you didn't use that here. It's not clear
>    to me why bird would be in its own module while the other algorithms remain
>    in the same module.
>    2. I'm not sure about the nrel.py module name. Do people frequently
>    associate this algorithm with NREL?
>    3. The bird function does a lot: calculates solar position, airmass,
>    atmospheric properties, ET irradiance, and finally clear sky. I think the
>    function and its signature should be simplified by requiring users to do
>    more work before they get here. This would also be consistent with the way
>    the existing clear sky algorithms work. So, I think the signature should be
>    more like
>
> bird(zenith, airmass_relative, aod700=0.1, precipitable_water=1.,
> ozone=0.3, pressure=101325., dni_extra=1364.)
>
> This is also related to the ineichen refactoring that I did in #199
> <https://github.com/pvlib/pvlib-python/pull/199>. The documentation can
> make note of the methods that need to be used on the input data in order to
> exactly reproduce the traditional model.
>
> Should we strip out the test data from the excel file and put it in a csv
> file? Advantages are that the csv file is smaller and we avoid the xlrd
> dependency. More importantly, are we allowed to redistribute the NREL file?
>
> Why add the requirements file? setup.py already has the minimum pvlib
> requirements in it.
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > +http://rredc.nrel.gov/solar/models/clearsky/
> +http://rredc.nrel.gov/solar/pubs/pdfs/tr-642-761.pdf
> +http://rredc.nrel.gov/solar/models/clearsky/error_reports.html
> +"""
> +
> +import numpy as np
> +import pandas as pd
> +import seaborn as sns
> +import xlrd
> +import os
> +
> +
> +def bird(doy, hr, lat, lon, tz, press_mB, o3_cm, h2o_cm, aod_500nm, aod_380nm,
> +         b_a, alb, lyear=365.0, solar_constant=1367.0):
> +    """
> +    Bird Simple Clearsky Model
>
> use numpydoc
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > +    )
> +    declination = np.rad2deg(dec_rad)
> +    # equation of time
> +    eqt = 229.18 * (
> +        0.0000075 + 0.001868 * np.cos(day_angle) -
> +        0.032077 * np.sin(day_angle) - 0.014615 * np.cos(2.0 * day_angle) -
> +        0.040849 * np.sin(2.0 * day_angle)
> +    )
> +    hour_angle = 15.0 * (hr - 12.5) + lon - tz * 15.0 + eqt / 4.0
> +    lat_rad = np.deg2rad(lat)
> +    ze_rad = np.arccos(
> +        np.cos(dec_rad) * np.cos(lat_rad) * np.cos(np.deg2rad(hour_angle)) +
> +        np.sin(dec_rad) * np.sin(lat_rad)
> +    )
> +    zenith = np.rad2deg(ze_rad)
> +    airmass = np.where(zenith < 89,
>
> This is the same as relativeairmass(method='kasten1966')
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > +                            for n, h in enumerate(headers)},
> +                            index=dt[1:48])
> +    assert np.allclose(testdata['Dangle'], day_angle[1:48])
> +    assert np.allclose(testdata['DEC'], declination[1:48])
> +    assert np.allclose(testdata['EQT'], eqt[1:48])
> +    assert np.allclose(testdata['Hour Angle'], hour_angle[1:48])
> +    assert np.allclose(testdata['Zenith Ang'], zenith[1:48])
> +    assert np.allclose(testdata['Air Mass'], airmass[1:48])
> +    assert np.allclose(testdata['Direct Beam'], Eb[1:48])
> +    assert np.allclose(testdata['Direct Hz'], Ebh[1:48])
> +    assert np.allclose(testdata['Global Hz'], Gh[1:48])
> +    assert np.allclose(testdata['Dif Hz'], Dh[1:48])
> +    return pd.DataFrame({'Eb': Eb, 'Ebh': Ebh, 'Gh': Gh, 'Dh': Dh}, index=dt)
> +
> +
> +if __name__ == "__main__":
>
> Hm, we don't have any script code in pvlib, but I would consider keeping
> this in a test module.
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > +    gh = np.where(airmass > 0, (id_nh + ias) / (1.0 - alb * rs), 0.0)
> +    testvalues = (day_angle, declination, eqt, hour_angle, zenith, airmass)
> +    return id_, id_nh, gh, gh - id_nh, testvalues
> +
> +
> +def etr(doy, lyear=365.0, solar_constant=1367.0):
> +    a0 = 1.00011
> +    doy0 = doy - 1.0
> +    a1 = 0.034221 * np.cos(2.0 * np.pi * doy0 / lyear)
> +    a2 = 0.00128 * np.sin(2.0 * np.pi * doy0 / lyear)
> +    a3 = 0.000719 * np.cos(2.0 * (2.0 * np.pi * doy0 / lyear))
> +    a4 = 0.000077 * np.sin(2.0 * (2.0 * np.pi * doy0 / lyear))
> +    return solar_constant * (a0 + a1 + a2 + a3 + a4)
> +
> +
> +def test_bird():
>
> tests should go in a module named test_[__name__].py. We should probably
> make a tests subsubpackage within each subpackage e.g.
> pvlib/clearsky/tests/test_bird.py
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > +        0.9662 * etr_ * t_aerosol * t_water * t_gases * t_ozone * t_rayliegh,
> +        0.0
> +    )
> +    id_nh = np.where(zenith < 90, id_ * np.cos(ze_rad), 0.0)
> +    ias = np.where(airmass > 0,
> +        etr_ * np.cos(ze_rad) * 0.79 * t_ozone * t_gases * t_water * taa *
> +        (0.5 * (1.0 - t_rayliegh) + b_a * (1.0 - (t_aerosol / taa))) / (
> +            1.0 - airmass + airmass ** 1.02
> +        ), 0.0
> +    )
> +    gh = np.where(airmass > 0, (id_nh + ias) / (1.0 - alb * rs), 0.0)
> +    testvalues = (day_angle, declination, eqt, hour_angle, zenith, airmass)
> +    return id_, id_nh, gh, gh - id_nh, testvalues
> +
> +
> +def etr(doy, lyear=365.0, solar_constant=1367.0):
>
> This the same as irradiance.extraradiation(method='spencer')
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > +        0.032077 * np.sin(day_angle) - 0.014615 * np.cos(2.0 * day_angle) -
> +        0.040849 * np.sin(2.0 * day_angle)
> +    )
> +    hour_angle = 15.0 * (hr - 12.5) + lon - tz * 15.0 + eqt / 4.0
> +    lat_rad = np.deg2rad(lat)
> +    ze_rad = np.arccos(
> +        np.cos(dec_rad) * np.cos(lat_rad) * np.cos(np.deg2rad(hour_angle)) +
> +        np.sin(dec_rad) * np.sin(lat_rad)
> +    )
> +    zenith = np.rad2deg(ze_rad)
> +    airmass = np.where(zenith < 89,
> +        1.0 / (np.cos(ze_rad) + 0.15 / (93.885 - zenith) ** 1.25), 0.0
> +    )
> +    pstar = press_mB / patm
> +    am_press = pstar*airmass
> +    t_rayliegh = np.where(airmass > 0,
>
> Are any of the np.where statements repetitive? In any case, calculate the airmass
> > 0 mask once and reuse it in subsequent tests.
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > +    )
> +    etr_ = etr(doy, lyear, solar_constant)
> +    id_ = np.where(airmass > 0,
> +        0.9662 * etr_ * t_aerosol * t_water * t_gases * t_ozone * t_rayliegh,
> +        0.0
> +    )
> +    id_nh = np.where(zenith < 90, id_ * np.cos(ze_rad), 0.0)
> +    ias = np.where(airmass > 0,
> +        etr_ * np.cos(ze_rad) * 0.79 * t_ozone * t_gases * t_water * taa *
> +        (0.5 * (1.0 - t_rayliegh) + b_a * (1.0 - (t_aerosol / taa))) / (
> +            1.0 - airmass + airmass ** 1.02
> +        ), 0.0
> +    )
> +    gh = np.where(airmass > 0, (id_nh + ias) / (1.0 - alb * rs), 0.0)
> +    testvalues = (day_angle, declination, eqt, hour_angle, zenith, airmass)
> +    return id_, id_nh, gh, gh - id_nh, testvalues
>
> Better to do the subtraction on a separate line.
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > @@ -0,0 +1,187 @@
> +"""
> +Richard E. Bird
> +        Clear Sky Broadband
> +        Solar Radiation Model
> +From the publication "A Simplified Clear Sky model for Direct and Diffuse
> +Insolation on Horizontal Surfaces" by R.E. Bird and R.L Hulstrom, SERI Technical
> +Report SERI/TR-642-761, Feb 1991. Solar Energy Research Institute, Golden, CO.
>
> 1981
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > @@ -0,0 +1,187 @@
> +"""
> +Richard E. Bird
> +        Clear Sky Broadband
> +        Solar Radiation Model
> +From the publication "A Simplified Clear Sky model for Direct and Diffuse
> +Insolation on Horizontal Surfaces" by R.E. Bird and R.L Hulstrom, SERI Technical
> +Report SERI/TR-642-761, Feb 1991. Solar Energy Research Institute, Golden, CO.
> +
> +The model is based on comparisons with results from rigourous radiative transfer
>
> text needs to be edited so that it's not specific to the excel file. Not
> sure if there is a licensing/attribution issue here.
> ------------------------------
>
> In pvlib/clearsky/nrel.py
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>
> :
>
> > +
> +http://rredc.nrel.gov/solar/models/clearsky/
> +http://rredc.nrel.gov/solar/pubs/pdfs/tr-642-761.pdf
> +http://rredc.nrel.gov/solar/models/clearsky/error_reports.html
> +"""
> +
> +import numpy as np
> +import pandas as pd
> +import seaborn as sns
> +import xlrd
> +import os
> +
> +
> +def bird(doy, hr, lat, lon, tz, press_mB, o3_cm, h2o_cm, aod_500nm, aod_380nm,
> +         b_a, alb, lyear=365.0, solar_constant=1367.0):
> +    """
>
> Some (all?) of the edited doc string above should be moved here.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pvlib/pvlib-python/pull/276#pullrequestreview-11283767>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH66AV_G_c7XNkhsM_6sTWTQPYQwK6waks5rEeKRgaJpZM4LDQh7>
> .
>

mikofski commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+    return pd.DataFrame({'Eb': Eb, 'Ebh': Ebh, 'Gh': Gh, 'Dh': Dh}, index=dt)
+
+
+if __name__ == "__main__":
```
removed script - it was just for debugging
:heavy_check_mark:

mikofski commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+    return solar_constant * (a0 + a1 + a2 + a3 + a4)
+
+
+def test_bird():
```
:heavy_check_mark:

mikofski commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+def bird(doy, hr, lat, lon, tz, press_mB, o3_cm, h2o_cm, aod_500nm, aod_380nm,
+         b_a, alb, lyear=365.0, solar_constant=1367.0):
+    """
+    Bird Simple Clearsky Model
```
:heavy_check_mark:

mikofski commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+        np.sin(dec_rad) * np.sin(lat_rad)
+    )
+    zenith = np.rad2deg(ze_rad)
+    airmass = np.where(zenith < 89,
```
* removed this calc in bird since passing relative airmass as argument
* use this in test_clearsky.test_bird()
:heavy_check_mark:

mikofski commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+    )
+    pstar = press_mB / patm
+    am_press = pstar*airmass
+    t_rayliegh = np.where(airmass > 0,
```
not necessary anymore, since relative airmass is `NaN` if not valid
:heavy_check_mark:

mikofski commented on pvlib/clearsky/nrel.py:
```diff
@@ -0,0 +1,187 @@
+    return id_, id_nh, gh, gh - id_nh, testvalues
+
+
+def etr(doy, lyear=365.0, solar_constant=1367.0):
```
using `dni_extra` arg now, so this is removed, but I'll use it in the test
:heavy_check_mark:

mikofski commented:
I think this is done. Some notes:

* I reversed the API changes. I'll leave that to you. Personally I like to use the dunder classes for breaking up modules into packages, but the `api.py` idea seems interesting too.
* I refactored the bird model to remove any redundant solarposition and atmosphere stuff, like you said. For testing I added some basic solar position stuff like analytical zenith, equation of time and declination
* I also refactored the signature to match other clearsky models
* if you really want to remove the xls bird file in lieu of a csv, that's fine, but it is a primary source, and I don't think there are any license issues, it is publicly available without registering, but guess we should check on that. I would think like 90% of users will already have xlrd, PITA for build testing huh
* the requirements file is really really useful!

    1. if you weren't using conda, travis would just look at your requirements and build your venv from that
    2. it makes it really easy to make a venv, which is what I've done for pvlib and why I made the file. Most projects recommend it, [EG see this](https://vladcalin.github.io/what-every-python-project-should-have.html#what-every-python-project-should-have)

* either python 3.4 or 3.5 randomly fails for no reason, something in test_forecast using netCDF4, but it doesn't always happen, and sometimes it changes from one version to another. What's up?

wholmgren commented:
Sorry again for the delay @mikofski. Does anyone have time to look at this more carefully? 

Without looking closely (and thus running a high chance of saying something stupid)...

> I reversed the API changes.

API looks good, thanks. I like the idea of reorganizing with api.py or something else, but let's do it in a separate PR. 

> I refactored the bird model to remove any redundant solarposition and atmosphere stuff, like you said. For testing I added some basic solar position stuff like analytical zenith, equation of time and declination

I was initially unhappy to see all of this less precise solar position functionality in a clear sky model PR. But I guess I'm ok with these functions if we document that users should prefer other functions for high precision calculations. Maybe put them in a different subsection in the API documentation? 

> if you really want to remove the xls bird file in lieu of a csv, that's fine, but it is a primary source, and I don't think there are any license issues, it is publicly available without registering, but guess we should check on that. I would think like 90% of users will already have xlrd, PITA for build testing huh

I like plain text for smaller data because people can view it on github, in diffs, and their editors. But this is fine.

> the requirements file is really really useful!

The base requirements are [numpy 1.9, pandas 0.14, pytz, and six](https://github.com/pvlib/pvlib-python/blob/master/setup.py#L43-L47). Maybe you want to make two requirements files -- requirements and requirements-all? Why the newer numpy scipy version limits?

> either python 3.4 or 3.5 randomly fails for no reason, something in test_forecast using netCDF4, but it doesn't always happen, and sometimes it changes from one version to another. What's up?

see #293. 

Are the keyword parameters sufficiently tested? Does it work with scalars?

mikofski commented:
still working on it. Any comments from anyone?

>Maybe put them in a different subsection in the API documentation

good idea, I'll put them in a section after spa with text prefixing that says these are low accuracy solar position methods

>I like plain text for smaller data because people can view it on github, in diffs, and their editors

remove xls, and reversed xlrd dependencies everywhere and use csv instead

>The base requirements are numpy 1.9, pandas 0.14, pytz, and six. Maybe you want to make two requirements files -- requirements and requirements-all? Why the newer numpy scipy version limits?

I removed the `requirements.txt`, too much trouble if no one cares but me, the version dependencies were accidental, but there were other dependencies to build sphinx for example, and ephem, etc, it's just not worth it

>Are the keyword parameters sufficiently tested? Does it work with scalars?

Nope, still need to check on this

mikofski commented:
@wholmgren and others:

This is ready for your review now:
1. moved low precision solar position correlations to separate section of documentation on `API.rst`
2. added keyword and scalar tests - passing

wholmgren commented:
Thanks, Mark. Might be next week before I can look carefully. Other users should feel free to chime in.

wholmgren commented on docs/sphinx/source/api.rst:
```diff
@@ -60,6 +61,19 @@ algorithm.
+   solarposition.solar_zenith_analytical
+   solarposition.declination_spencer71
+   solarposition.declination_cooper69
+   solarposition.equation_of_time_Spencer71
```
lower case s here and in function definition.

mikofski replied:
:ballot_box_with_check:

wholmgren commented on pvlib/clearsky.py:
```diff
@@ -720,3 +720,103 @@ def rmse(alpha):
+    Parameters
+    ----------
+    zenith : numeric
+        Solar zenith angle in degrees.
```
solar_zenith or apparent_zenith? http://pvlib-python.readthedocs.io/en/latest/variables_style_rules.html

mikofski replied:
`solar_zenith` not apparent - in case I forget

wholmgren commented on pvlib/clearsky.py:
```diff
@@ -720,3 +720,103 @@ def rmse(alpha):
+    NREL Bird Simple Clear Sky Broadband Solar Radiation Model
+
+    Based on original implementation by Daryl R. Myers at NREL.
+
```
You could add some notes about how zenith and airmass should be calculated if users want to exactly reproduce the entire bird model.

wholmgren replied:
Or this could go into a new section in the clearsky.rst documentation.

wholmgren commented on pvlib/clearsky.py:
```diff
@@ -720,3 +720,103 @@ def rmse(alpha):
+
+    Returns
+    -------
+    direct_beam : numeric
```
It's more conventional in pvlib to return a DataFrame or OrderedDict, rather than a tuple. Series in, DataFrame out, anything else in, OrderedDict out. The other functions in this module demonstrate this pattern. columns/keys should follow http://pvlib-python.readthedocs.io/en/latest/variables_style_rules.html

mikofski commented:
@wholmgren sorry for being so slow. I am still working on this.

mikofski commented:
@wholmgren and @cwhanse I think this is finally ready to be merged. It passes all tests. I've addressed the return types by converting them to `OrderdDict` or `DataFrame` and ordered the output similar to other clear sky models. I did not address the concerns about adding documentation to `clearsky.rst` about the inputs to duplicate the Daryl Myers Excel spreadsheet, or add any new significant documentation other that to the API. Perhaps I can address additional documentation as needed in a future realease? Thanks!

mikofski commented:
@wholmgren I made minor documentation changes to address the testing parameters and solar or apparent zenith argument. Please let me know if there are any other issues to address before you can merge this. I would like to use the Bird model for comparisons of AOD/Pwat data with Linke turbidity in with Ineichen's original and newer simplified-Solis clearsky model for our upcoming PVSC paper. :wink:

wholmgren commented:
That all sounds good, but I probably will not be able to look at this
closely until next wednesday.

On Thu, May 18, 2017 at 2:44 PM Mark Mikofski <notifications@github.com>
wrote:

> @wholmgren <https://github.com/wholmgren> I made minor documentation
> changes to address the testing parameters and solar or apparent zenith
> argument. Please let me know if there are any other issues to address
> before you can merge this. I would like to use the Bird model for
> comparisons of AOD/Pwat data with Linke turbidity in with Ineichen's
> original and newer simplified-Solis clearsky model for our upcoming PVSC
> paper. 😉
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pvlib/pvlib-python/pull/276#issuecomment-302550433>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AELiR2G6PWT34mBjPNJkVjl1eRUMAWF1ks5r7LvTgaJpZM4LDQh7>
> .
>

wholmgren commented on pvlib/test/test_solarposition.py:
```diff
@@ -367,3 +367,15 @@ def test_nrel_earthsun_distance():
     expected = pd.Series(np.array([0.983289204601]),
                          index=pd.DatetimeIndex([times, ]))
     assert_series_equal(expected, result)
+
```
shouldn't we add tests for all of the other functions you've added to ``solarposition.py``? I think they should be marked private if you don't want to add the tests.

mikofski replied:
Added 2 more tests for low accuracy declination and solar-zenith formulas.

mikofski commented:
@wholmgren the test files are each 360-kb, so total combined less than 1-mb, okay?

![screenshot from 2017-05-24 20-32-45](https://cloud.githubusercontent.com/assets/1385621/26434897/4feff496-40c0-11e7-9e24-cef713c81ed6.png)

wholmgren commented:
ok on the test file size. 

I think this is ready to go. Agreed?

mikofski commented:
Unless anyone else has any comments, _yes_!
